frontend/test/resolution/testGetSymbolsAvailableInScope.cpp

```diff
@@ -0,0 +1,195 @@
+
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/resolution/resolution-queries.h"
+#include "chpl/resolution/scope-queries.h"
+#include "chpl/uast/Block.h"
+#include "chpl/uast/Identifier.h"
+#include "chpl/uast/Include.h"
+#include "chpl/uast/Module.h"
+#include "chpl/uast/Variable.h"
+
+// helper functions
+
+struct NameFinder {
+  std::set<UniqueString> names;
+
+  bool enter(const NamedDecl* named) {
+    names.insert(named->name());
+    return true;
+  }
+
+  bool enter(const Identifier* ident) {
+    names.insert(ident->name());
+    return true;
+  }
+
+  bool enter(const AstNode* node) {
+    return true;
+  }
+
+  void exit(const AstNode* node) {}
+};
+
+static void checkScopeContents(Context* context, const Module* mod,
+                               std::vector<std::string>&& expected) {
+  auto scope = scopeForId(context, mod->id());
+
+  std::vector<std::string> actual;
+  for (auto& symPair : getSymbolsAvailableInScope(context, scope)) {
+    actual.push_back(symPair.first.str());
+  }
+  std::sort(actual.begin(), actual.end());
+  std::sort(expected.begin(), expected.end());
+
+  assert(actual.size() == expected.size());
+  for (size_t i = 0; i < expected.size(); i++) {
+    assert(actual[i] == expected[i]);
+  }
+}
+
+static void checkScopeContentsViaLookup(Context* context, const Module* mod,
+                                        const std::set<UniqueString>& allNames) {
+  // Verify that looking up identifiers in the scope and asking for
+  // all identifiers matches up.
+  //
+  // The property should be:
+  //
+  //   for each symbol s,
+  //     lookup(scope, s) nonempty iff s in getSymbolsAvailableInScope(scope)
+
+  auto scope = scopeForId(context, mod->id());
+  auto allSyms = getSymbolsAvailableInScope(context, scope);
+
+  for (auto name : allNames) {
+    // 'int' is visible from all scopes, but we don't list it in allSyms.
+    if (name == USTR("int")) continue;
+
+    auto result = lookupNameInScope(context, scope, {}, name, IDENTIFIER_LOOKUP_CONFIG);
+
+    if (result.empty()) {
+      assert(allSyms.find(name) == allSyms.end());
+    } else {
+      assert(allSyms.find(name) != allSyms.end());
+    }
+  }
+}
+
+static void test1() {
+  printf("test1\n");
+  Context ctx;
+  Context* context = &ctx;
+
+  auto path = UniqueString::get(context, "input.chpl");
+  std::string contents = R""""(
+      module M {
+        var x: int;
+
+        private proc f1() {}
+        proc f2() {}
+      }
+      module N {
+        var y: int;
+
+        private proc f3() {}
+        proc f4() {}
+      }
+      module O {
+        var z: int;
+
+        private proc f5() {}
+        proc f6() {}
+      }
+      module P {
+        public use M;
+        use N;
+        public use O as O;
+      }
+      module Q {
+        use P;
+      }
+   )"""";
+  setFileText(context, path, contents);
+
+  const ModuleVec& vec = parseToplevel(context, path);
+
+  checkScopeContents(context, vec[0], {"x", "f1", "f2"});
```

Comment:
Arguably, it should also include `M` (but oddly enough, only for use in `use` / `import`). Also, arguably, it should include the included-by-default symbols. It's not important to me that it change behavior from what we have right now, but might be worth a comment here just to make updating the test easier.